src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/SharedScope.java

```diff
@@ -184,6 +184,7 @@ public void close() {
                         throw new IllegalStateException("Already closed");
                     }
                 } while (!STATE.compareAndSet(jdk.internal.foreign.SharedScope.this, value, value - 1));
+                Reference.reachabilityFence(SharedScope.this);
```

Comment:
Not sure this is needed - I added it as we were worried that possible optimization (e.g. scalarization) might break reachability (this class is an innner class and keeps a strong reference to its enclosing, after all).